plugins/plugin-machine/che-plugin-machine-ext-client/src/main/java/org/eclipse/che/ide/extension/machine/client/actions/SelectCommandComboBoxReady.java

```diff
@@ -333,17 +315,141 @@ public void onConfigurationsUpdated(CommandConfiguration command) {
     @Override
     public void onWsAgentStarted(WsAgentStateEvent event) {
         loadCommands(null);
+        loadMachines();
     }
 
+    private void loadMachines() {
+        machineServiceClient.getMachines(workspaceId).then(new Operation<List<MachineDto>>() {
+            @Override
+            public void apply(List<MachineDto> machines) throws OperationException {
+                addMachineActions(machines);
+            }
+        });
+    }
+
+
     @Override
     public void onWsAgentStopped(WsAgentStateEvent event) {
+    }
+
+    @Override
+    public void onMachineRunning(MachineStateEvent event) {
+        MachineDto machine = event.getMachine();
 
+        addMachineAction(machine);
     }
 
-    private static class CommandsComparator implements Comparator<CommandConfiguration> {
+    @Override
+    public void onMachineDestroyed(MachineStateEvent event) {
+        MachineDto machine = event.getMachine();
+
+        this.removeMachineAction(machine);
+    }
+
+    private void addMachineActions(List<MachineDto> machines) {
+        for (MachineDto machine : machines) {
+            registeredMachineMap.put(machine.getId(), machine);
+        }
+        this.updateMachineActions();
+    }
+
+    private void addMachineAction(MachineDto machine) {
+        registeredMachineMap.put(machine.getId(), machine);
+        this.updateMachineActions();
+
+        machinesListWidget.selectElement(machine.getId(), machine.getConfig().getName());
+    }
+
+    private boolean removeMachineAction(MachineDto machine) {
```

Comment:
Moreover the method id user once inside the class in
org.eclipse.che.ide.extension.machine.client.actions.SelectCommandComboBox#onMachineDestroyed
So the code is combined.